modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+        Resource.liftF(settings.valueSerializer).flatMap { valueSerializer =>
+          settings.blocker.fold(Blockers.producer)(Resource.pure[F, Blocker]).flatMap { blocker =>
+            buildGrouper(blocker).map { grouper =>
+              new TransactionalKafkaProducer[F, K, V] {
+                override def produce[P](
+                  records: TransactionalProducerRecords[F, K, V, P]
+                ): F[ProducerResult[Chunk, K, V, P]] =
+                  grouper(records.records)
+                    .flatMap { recordGroups =>
+                      recordGroups.flatTraverse {
+                        case (producer, recordGroup) =>
+                          produceTransaction(producer, recordGroup)
+                      }
+                    }
+                    .map(ProducerResult(_, records.passthrough))
 
-                  consumerGroupId.flatMap { groupId =>
-                    withProducer { producer =>
-                      F.bracketCase(F.delay(producer.beginTransaction())) { _ =>
-                        val producerRecords: ArrayBuffer[ProducerRecord[K, V]] =
-                          new ArrayBuffer(records.records.size)
+                private[this] def produceTransaction(
+                  producer: KafkaByteProducer,
+                  records: RecordsChunk
+                ): F[Chunk[(ProducerRecord[K, V], RecordMetadata)]] =
+                  if (records.isEmpty) {
+                    F.pure(Chunk.empty)
+                  } else {
+                    val batch = CommittableOffsetBatch.fromFoldableMap(records)(_.offset)
 
-                        records.records.foreach { committable =>
-                          committable.foldable.foldLeft(committable.records, ()) {
-                            case (_, record) =>
-                              producerRecords += record
-                              ()
-                          }
-                        }
+                    val consumerGroupId =
+                      if (batch.consumerGroupIdsMissing || batch.consumerGroupIds.size != 1) {
+                        F.raiseError[String](ConsumerGroupException(batch.consumerGroupIds))
+                      } else {
+                        F.pure(batch.consumerGroupIds.head)
+                      }
 
-                        Chunk
-                          .buffer(producerRecords)
-                          .traverse(
-                            KafkaProducer.produceRecord(keySerializer, valueSerializer, producer)
-                          )
-                          .map(_.sequence)
-                          .flatTap { _ =>
-                            F.delay {
-                              producer.sendOffsetsToTransaction(
-                                batch.offsets.asJava,
-                                groupId
-                              )
+                    consumerGroupId.flatMap { groupId =>
+                      records.flatTraverse { committable =>
```

Comment:
This is one block where replacing `G[_]` with `Chunk` pays off, because it makes interleaving the various `traverse` + `flatten` combinators much easier.